### ISSUE TITLE
PMML for categorical attributes displays in DataDictionary the _mapped  attributes as well.

### ISSUE THREAD

**Issue by radescu** [catboost/catboost#1025]
Problem: 
The PMML from catboost  (using BinaryClassification) is defining a new input attribute for every categorical attribute for internal use (the one that was mapped from string to integers).  This definition of DataField should not be done under <DataDictionary> but under <DerivedField> in the PMML structure.

The problem is the way it is now that when importing the PMML into a software application (the application that calls this PMML) a user then needs to define this new input attribute for every categorical attribute, thus doubling the number of inputs for categorical attributes.  Although this is better than defining a new attribute for each category  (the explosion of features as done by one-hot-encoding), it's still not optimally done since it still requires doubling of categorical attributes in the input definition.
Expected behaviour is that <DataDictionary> does not list the "internally" transformed attributes,  but  these are listed under <DerivedField> where the mapping is applied.

I attach the screenshot of the PMML using current behaviour:
![image](https://user-images.githubusercontent.com/12171505/66325258-3729e680-e927-11e9-9c7c-41e701c74b3c.png)
The L_CatMcc is the original categorical attribute,
the L_CatMcc_mapped should not be under DataDictionary.

Expected behaviour:
![image](https://user-images.githubusercontent.com/12171505/66326184-db605d00-e928-11e9-90d3-5dee5f805bac.png)
with the Value mapped in the DerivedField.

catboost version:0.15.2
    Commit: e9f5a090056f213b7b6e0657a4e919932bccdf34
    Branch: heads/master
Operating System: macOS 10.14.5 (python environment 3.7)
PMML: 4.3
CPU:
# GPU:

**Comment by annaveronika** [comment]
We agree with your proposal. It would be really great if you would help us with this change. Can you do that?

**Comment by radescu** [comment]
could you please point me to the code - i'll take a look.

**Comment by andrey-khropov** [comment]
> could you please point me to the code - i'll take a look.

Code to export trained models is contained in [catboost/libs/model/model_export/pmml_export.{h|cpp} files](https://github.com/catboost/catboost/blob/master/catboost/libs/model/model_export/pmml_helpers.h).

Code to output data dictionary is [here](https://github.com/catboost/catboost/blob/708a78c30c264bf73fd5b1a27256a97b32bb494e/catboost/libs/model/model_export/pmml_helpers.cpp#L68).

There are regression tests that check that the pmml export output is canonical, they are [here](https://github.com/catboost/catboost/blob/708a78c30c264bf73fd5b1a27256a97b32bb494e/catboost/python-package/ut/medium/test.py#L1112).
All datasets in these tests contain categorical features.
Instructions how to run tests are located [here](https://catboost.ai/docs/concepts/development-and-contributions.html#run-tests)

Currently, there are no tests to check that exported PMML model produces expected results. 
The validity of current results has been checked by comparing predictions from CatBoost native model evaluator (in python package it's [predict](https://catboost.ai/docs/concepts/python-reference_catboost_predict.html) function) with some model evaluator that supports PMML model.

I used the following code with the [jpmml evaluator](https://github.com/jpmml/jpmml-evaluator) (model.pmml was taken [from catboost tests](https://github.com/catboost/catboost/blob/master/catboost/python-package/ut/medium/canondata/test.test_pmml_export_binclass_/model.pmml) ):

```java
package com.mycompany.app;
 
import java.io.*;
import java.util.*;
 
import org.xml.sax.SAXException;
 
import org.dmg.pmml.*;
import org.jpmml.model.*;
import org.jpmml.evaluator.*;
 
public class App 
{
    public static void main(String[] args) throws Exception
    {
        String modelPath = "./model.pmml";
        
        Evaluator evaluator = new LoadingModelEvaluatorBuilder()
            .setLocatable(false)
            .setVisitors(new DefaultVisitorBattery())
            .load(new File(modelPath))
            .build();
    
        Map<String, Object> inputDataRecord = new HashMap<String,Object>();
        inputDataRecord.put("feature_0", "a");
        inputDataRecord.put("feature_1", "a");
        inputDataRecord.put("feature_2", "a");
        inputDataRecord.put("feature_3", 0.1);
        inputDataRecord.put("feature_4", "Federal-gov");
        inputDataRecord.put("feature_5", 0.0);
        inputDataRecord.put("feature_6", "a");
        inputDataRecord.put("feature_7", 1.2);
        inputDataRecord.put("feature_8", "Married-spouse-absent");
        inputDataRecord.put("feature_9", "Prof-specialty");
        inputDataRecord.put("feature_10", "Husband");
        inputDataRecord.put("feature_11", "a");
        inputDataRecord.put("feature_12", "a");
        inputDataRecord.put("feature_12", "a");
        inputDataRecord.put("feature_13", 12.0);
        inputDataRecord.put("feature_14", 0.12);
        inputDataRecord.put("feature_15", 0.62);
        inputDataRecord.put("feature_16", "Canada");
        
        Map<FieldName, FieldValue> arguments = new LinkedHashMap<>();
        
        List<? extends InputField> inputFields = evaluator.getInputFields();
        for(InputField inputField : inputFields){
            FieldName inputName = inputField.getName();
            
            Object rawValue = inputDataRecord.get(inputName.getValue());
            
            // Transforming an arbitrary user-supplied value to a known-good PMML value
            // The user-supplied value is passed through: 1) outlier treatment, 2) missing value treatment, 3) invalid value treatment and 4) type conversion
            FieldValue inputValue = inputField.prepare(rawValue);
            
            arguments.put(inputName, inputValue);
        }
    
        Map<FieldName, ?> results = evaluator.evaluate(arguments);
    
        List<? extends TargetField> targetFields = evaluator.getTargetFields();
        for(TargetField targetField : targetFields){
            FieldName targetName = targetField.getName();
            
            Object targetValue = results.get(targetName);
            System.out.println(targetName);
            System.out.println(targetValue);
        }
        
    }
```

You should check that after your changes evaluator produces the same predictions as before.

**Comment by annaveronika** [comment]
@radescu Was this helpful? Do you need any additional information?

**Comment by radescu** [comment]
@annaveronika @andrey-khropov thank you, unfortunately I had no time to look and I wont be able to provide more feedback till middle of November.

**Comment by andrey-khropov** [comment]
This issue has been fixed by commit cc170dc1a0ee4461dfaeeb0fb6dfed33847d5ebb which will be included in the next release.
